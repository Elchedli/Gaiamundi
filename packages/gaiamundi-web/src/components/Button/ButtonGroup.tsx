import classNames from 'classnames';
import {
  Children,
  cloneElement,
  FC,
  PropsWithChildren,
  ReactElement,
  useMemo,
} from 'react';

import { ButtonComponentProps } from './Button';

export type ButtonGroupProps = PropsWithChildren<{
  pill?: boolean;
  outline?: boolean;
  className?: string;
}>;

const ButtonGroup: FC<ButtonGroupProps> = ({
  children,
  pill,
  outline,
  className,
}) => {
  const items = useMemo(
    () =>
      Children.map(
        children as ReactElement<ButtonComponentProps>[],
        (child, index) =>
          cloneElement(child, {
            pill,
            outline,
            positionInGroup:
              index === 0
                ? 'start'
                : index ===
                  (children as ReactElement<ButtonComponentProps>[]).length - 1
                ? 'end'
                : 'middle',
          })
      ),
    [children, outline, pill]
  );

  return (
    <div className={classNames('inline-flex', className)} role="group">
      {items}
    </div>
  );
};

ButtonGroup.displayName = 'Button.Group';
export default ButtonGroup;
