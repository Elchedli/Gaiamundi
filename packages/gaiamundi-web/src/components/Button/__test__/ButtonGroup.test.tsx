import { render } from '@testing-library/react';

import { Button } from 'components/Button/Button';
import ButtonGroup from 'components/Button/ButtonGroup';

describe('ButtonGroup', () => {
  it('renders correct classes', () => {
    const { getByRole } = render(<ButtonGroup />);
    const buttonGroup = getByRole('group');
    expect(buttonGroup).toHaveClass('inline-flex');
  });

  it('should render children', () => {
    const { container, getByText } = render(
      <ButtonGroup>
        <Button>Button 1</Button>
        <Button>Button 2</Button>
      </ButtonGroup>
    );
    expect(getByText('Button 1')).toBeInTheDocument();
    expect(getByText('Button 2')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should add the "pill | outline" class if the "pill | outline" prop is true', () => {
    const { container } = render(
      <ButtonGroup pill outline>
        <Button>Button 1</Button>
        <Button>Button 2</Button>
        <Button>Button 3</Button>
      </ButtonGroup>
    );
    expect(container.querySelector('span')).toHaveClass('rounded-full');
    expect(container).toMatchSnapshot();
  });

  it('should add the custom class if added with otherProps', () => {
    const { container } = render(
      <ButtonGroup className="testClass">
        <Button>Button 1</Button>
        <Button>Button 2</Button>
      </ButtonGroup>
    );
    expect(container.firstChild).toHaveClass('testClass');
    expect(container).toMatchSnapshot();
  });
});
