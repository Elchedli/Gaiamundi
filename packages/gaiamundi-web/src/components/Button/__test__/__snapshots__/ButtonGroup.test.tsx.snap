// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should add the "pill | outline" class if the "pill | outline" prop is true 1`] = `
<div>
  <div
    class="inline-flex"
    role="group"
  >
    <button
      class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-full text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 focus:!ring-2 rounded-r-none"
      data-testid="button-element"
      type="button"
    >
      <span
        class="flex items-center text-sm px-4 py-2 rounded-full bg-white text-gray-900 transition-all duration-75 ease-in group-hover:bg-opacity-0 group-hover:text-inherit dark:bg-gray-900 dark:text-white rounded-r-none"
      >
        Button 1
      </span>
    </button>
    <button
      class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-full text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 focus:!ring-2 !rounded-none border-l-0 pl-0"
      data-testid="button-element"
      type="button"
    >
      <span
        class="flex items-center text-sm px-4 py-2 rounded-full bg-white text-gray-900 transition-all duration-75 ease-in group-hover:bg-opacity-0 group-hover:text-inherit dark:bg-gray-900 dark:text-white !rounded-none"
      >
        Button 2
      </span>
    </button>
    <button
      class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-full text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 focus:!ring-2 rounded-l-none border-l-0 pl-0"
      data-testid="button-element"
      type="button"
    >
      <span
        class="flex items-center text-sm px-4 py-2 rounded-full bg-white text-gray-900 transition-all duration-75 ease-in group-hover:bg-opacity-0 group-hover:text-inherit dark:bg-gray-900 dark:text-white rounded-l-none"
      >
        Button 3
      </span>
    </button>
  </div>
</div>
`;

exports[`ButtonGroup should add the custom class if added with otherProps 1`] = `
<div>
  <div
    class="inline-flex testClass"
    role="group"
  >
    <button
      class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 focus:!ring-2 rounded-r-none"
      data-testid="button-element"
      type="button"
    >
      <span
        class="flex items-center text-sm px-4 py-2 rounded-md rounded-r-none"
      >
        Button 1
      </span>
    </button>
    <button
      class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 focus:!ring-2 rounded-l-none border-l-0 pl-0"
      data-testid="button-element"
      type="button"
    >
      <span
        class="flex items-center text-sm px-4 py-2 rounded-md rounded-l-none"
      >
        Button 2
      </span>
    </button>
  </div>
</div>
`;

exports[`ButtonGroup should render children 1`] = `
<div>
  <div
    class="inline-flex"
    role="group"
  >
    <button
      class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 focus:!ring-2 rounded-r-none"
      data-testid="button-element"
      type="button"
    >
      <span
        class="flex items-center text-sm px-4 py-2 rounded-md rounded-r-none"
      >
        Button 1
      </span>
    </button>
    <button
      class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 focus:!ring-2 rounded-l-none border-l-0 pl-0"
      data-testid="button-element"
      type="button"
    >
      <span
        class="flex items-center text-sm px-4 py-2 rounded-md rounded-l-none"
      >
        Button 2
      </span>
    </button>
  </div>
</div>
`;
